### PR TITLE
DIG-772: GraphQL Integration with CanDIGv2 stack - Part 2

### PR DESCRIPTION
**Pull request by AliRZ-02**
## Description
This PR is the second PR associated with the [DIG-772](https://candig.atlassian.net/browse/DIG-772) ticket, aiming to integrate the GraphQL Interface with the CanDIGv2 stack. This PR in particular deals more deeply into the CanDIG stack to integrate the pieces of the stack with the GraphQL interface. In general, the PR adds the GraphQL interface as a submodule and subdirectory to its existing stack, while modifying the authx services to accept requests to and from the GraphQL-interface. The interface is treated as a frontend module, due to its GraphiQL display. The PR does work to ensure that incoming requests pass through the Tyk service, auth headers are passed to their respective services, also through their respective Tyk endpoints. The environment variables are also modified in accordance with the CanDIGv2 stack, to ensure that the GraphQL interface can be set up with ease, and can also be easily modified.

## Changelog
- Added `graphql` subdirectory in `lib` containing a submodule called `GraphQL-interface`
- Modified Tyk Config to allow for incoming requests to `/graphql` endpoint
- Modified `etc/env/example.env` to include config setup for GraphQL-interface + Tyk GraphQL endpoint
- Wired up GraphQL-interface to Katsu and CanDIG services through .env 
- Added docker-compose.yml file to set up graphql service using `make` commands

## Dependencies
- This PR depends on [Part 1](https://github.com/CanDIG/GraphQL-interface/pull/11) of the CanDIGv2 GraphQL integration ticket, DIG-772

## Testing
- To build-test this PR, simply ensure that you have the following installed on your machine:
    - `git, docker, docker-compose, make`
- Follow the steps listed on the CanDIGv2 [docs](https://github.com/CanDIG/CanDIGv2/blob/develop/docs/install-docker.md) to install the candigv2 stack. Note that we will not need to install all of the modules in the stack, given the nature of the GraphQL-interface, and so please stop right after the `make init-docker` command.
- Then edit the `.env` file, modifying the `KEYCLOAK_GENERATE_TEST_USER` variable to be `1`. Also open up your `/etc/hosts` file and add the following lines:
    - ```bash
          <your_local_ipv4 address> docker.localhost
          <your_local_ipv4_address> auth.docker.localhost
- Finally, perform the following commands:
    - ```bash
         make build-chord-metadata
         make build-candig-server
         make build-graphql
         make init-authx
         make compose-chord-metadata
         make compose-candig-server
         make compose-graphql
- Then you should have the CanDIG v2 stack up and running with GraphQL also set. You can play around with the features and see if you can log in (the username and password are in `tmp/secrets/keycloak-test-user` and `tmp/secrets/keycloak-test-user-password` respectively. 

### With Testing Data
- If you also want to add testing data, you can modify the `candig-server` Dockerfile to add testing variant data, such as the one present in the Candigv1 server [docs](https://candig-server.readthedocs.io/en/v1.5.0-alpha/data.html#reads-variants-and-references-data). 
    - Add the following code to your `candig-server` Dockerfile, just before the last line (`ENTRYPOINT ["candig_server"]`), and then rebuild the `candig-server` container with `make build-candig-server && make compose-candig-server`:
```Dockerfile
RUN wget http://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/hg19.fa.gz && \
        gunzip hg19.fa.gz && \
        wget https://github.com/CanDIG/test_data/releases/download/v1.0.0/group1.tar && \
        wget https://github.com/CanDIG/test_data/releases/download/v1.0.0/group1_load.sh && \
        wget https://github.com/CanDIG/test_data/releases/download/v1.0.0/group1_clinphen.json

RUN tar xvf group1.tar && \
        sed -i 's/GRCh37-lite/hg19a/' group1_load.sh && \
        sed -i 's/registry.db/candig-example-data\/registry.db/' group1_load.sh && \
        chmod +x group1_load.sh && \
        ingest candig-example-data/registry.db test300 group1_clinphen.json && \
        candig_repo add-referenceset candig-example-data/registry.db hg19.fa -d "hg19a reference genome" --name hg19a && \
        bash ./group1_load.sh
```
- You can then follow the instructions in the GraphQL-interface subdirectory's [Data ingestion file](https://github.com/CanDIG/GraphQL-interface/blob/master/docs/GraphQLDataCollection.md#variant-patient-data-collection--ingestion) to add sample data to your katsu db. Make sure to modify the host and ports in your `GraphQL-interface/docs/helpers/CanDig-Katsu-Ingestion/defaults.py` folder to fit with those of the CanDIGv2 stack (look at `docker ps` to find the required ports and hosts). Note that the `projects` directory simply refers to a directory outside of the graphql subdirectory, you could put the `venv` in your CanDIGv2 root directory f you so choose.
- Then test with the following table by going to the GraphiQL UI, and inputting the following query, making sure to adjust the input values to those seen below:

```graphql
{
  beaconQuery(input: {referenceName: "CHR", start: STARTINGPOSITION, end: ENDINGPOSITION, referenceBases: "REFERENCE BASE", alternateBases: "ALTERNATE BASE"}) {
    exists
    error {
      errorCode
      errorMessage
    }
    alleleRequest {
      referenceName
      referenceBases
      start
      end
      alternateBases
    }
    apiVersion
    individualsPresent {
      personalInfo {
        sex
        ethnicity
        race
      }
      mcodepackets {
        cancerCondition {
          clinicalStatus {
            label
          }
          dateOfDiagnosis
        }
      }
      phenopackets {
        biosamples {
          description
        }
      }
    }
  }
}
```

| ID | CHR  | STARTING POSITION | ENDING POSITION | REFERENCE BASE | ALTERNATE BASE |
| --  | ----- | ----------------------- | --------------------- | ------------------- | ------------------- |
| 1   | 1      | 712800                       | 712900                    | T                           | C                          |
| 2   | 1      | 945100                       | 945200                    | C                           | T                          |
| 3   | 1      | 1368223                     | 1368225                  | GA                        | G                          |
| 4   | 1      | 19598000                   | 19598100                | AAAAAAAAG        | A                          |
| 5   | 2      | 121221                       | 121222                    | A                           | G                          |
| 6   | 2      | 32260                         | 32270                      | T                           | C                           |
| 7   | 7      | 84968                         | 84969                      | C                           | T                           |
| 8   | 9      | 231000                       | 232000                    | A                           | AT                         |
| 9   | 11    | 99925898                   | 99925899                | A                           | C                           |
| 10 | 13    | 19077200                   | 19077300                | G                           | A                           |
| 11 | 14    | 19456897                   | 19456898                | C                           | T                            |
| 12 | 15    | 20405300                   | 20405310                | T                           | A                            |
| 13 | 16    | 69848771                   | 69848772                | A                           | G                           |
| 14 | 17    | 827                             | 828                          | T                           | C                            |
| 15 | 17    | 14673964                   | 14673965                | G                           | A                           |
| 16 | 3      | 1                                 | 2                              | T                           | C                           |
| 17 | 5      | 5                                 | 14                            | A                           | G                           |
| 18 | 12    | 1234355                     | 1234365                  | ACTG                     | GTCA                     |
| 19 | 12    | 1873000                     | 1874000                  | C                           | T                            |
| 20 | 17    | 91023                         | 91024                      | T                           | C                            |

### CODE REVIEW

File: etc/env/example.env [hunk 1]
```diff
@@ -314,3 +321,16 @@ CANCOGEN_FEDERATION_URL=http://federation-service:4232/federation/search
 CANDIG_DATA_PORTAL_VERSION=v0.1.0
 CANDIG_DATA_PORTAL_PORT=2543
 CANDIG_DATA_PORTAL_URL=http://${CANDIG_DOMAIN}:${CANDIG_DATA_PORT}/data-portal
+
+# graphql-interface
+GRAPHQL_CONDA_SOURCE=condaforge
```
Comment (shaikh-rashid) [inline etc/env/example.env]: why does the container require conda?

File: etc/env/example.env [hunk 2]
```diff
@@ -314,3 +321,16 @@ CANCOGEN_FEDERATION_URL=http://federation-service:4232/federation/search
...
+GRAPHQL_CONDA_TYPE=miniforge3
+GRAPHQL_CONDA_VERSION=latest
+GRAPHQL_INTERFACE_VERSION=v1.0.0
+GRAPHQL_INTERFACE_PORT=7999
+
+GRAPHQL_KATSU_API=http://candigv2_tyk_1:8080/katsu/api
+GRAPHQL_CANDIG_SERVER=http://candigv2_tyk_1:8080/candig
+GRAPHQL_BEACON_ID=com.candig.graphql
+GRAPHQL_KATSU_TOKEN_KEY=authorization
```
Comment (shaikh-rashid) [inline etc/env/example.env:333]: Please see request for changes in https://github.com/CanDIG/GraphQL-interface/pull/11 regarding file secrets

File: lib/graphql/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+version: '3.7'
+
+services:
+  graphql:
+    build:
+      context: $PWD/lib/graphql/GraphQL-interface
+      args:
+        GRAPHQL_CONDA_SOURCE: "${GRAPHQL_CONDA_SOURCE}"
```
Comment (shaikh-rashid) [inline lib/graphql/docker-compose.yml]: After review of https://github.com/CanDIG/GraphQL-interface/blob/AliRZ-02/DIG-772-CanDIGv2-Integration/Dockerfile, it seems like Conda is being used unnecessarily. Conda is not needed when deploying in a container. The `scikitlearn` and `pandas` dependencies can be installed on host with `pip`. Additionally, base image in not recommended as it is adds unnecessary overhead.

File: lib/graphql/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        - "traefik.http.routers.graphql.rule=Host(`graphql.${CANDIG_DOMAIN}`)"
+        - "traefik.http.routers.graphql.entrypoints=${TRAEFIK_ENTRYPOINT}"
+        - "traefik.http.services.graphql.loadbalancer.server.port=${GRAPHQL_INTERFACE_PORT}"
+    logging: *default-logging
+    environment:
+      - GRAPHQL_KATSU_API=${GRAPHQL_KATSU_API}
+      - GRAPHQL_CANDIG_SERVER=${GRAPHQL_CANDIG_SERVER}
+      - GRAPHQL_BEACON_ID=${GRAPHQL_BEACON_ID}
+      - GRAPHQL_KATSU_TOKEN_KEY=${GRAPHQL_KATSU_TOKEN_KEY}
+      - GRAPHQL_CANDIG_TOKEN_KEY=${GRAPHQL_CANDIG_TOKEN_KEY}
```
Comment (shaikh-rashid) [inline lib/graphql/docker-compose.yml:35]: Please see previous comments regarding file secrets.

**Review by shaikh-rashid — CHANGES_REQUESTED** [review]
Great work, minor changes requested

File: etc/env/example.env (hunk 2 above)
Comment (AliRZ-02) [reply etc/env/example.env:333]: Fixed: See https://github.com/CanDIG/GraphQL-interface/pull/11#discussion_r819079892

File: lib/graphql/docker-compose.yml (hunk 4 above)
Comment (AliRZ-02) [reply lib/graphql/docker-compose.yml:35]: Fixed: See https://github.com/CanDIG/GraphQL-interface/pull/11#discussion_r819079892

File: etc/env/example.env (hunk 1 above)
Comment (AliRZ-02) [reply etc/env/example.env]: Fixed: See https://github.com/CanDIG/CanDIGv2/pull/118

File: lib/tyk/configuration_templates/api_graphql.json.tpl [hunk 5]
```diff
@@ -0,0 +1,153 @@
+{
```
Comment (Zamm178) [inline lib/tyk/configuration_templates/api_graphql.json.tpl:1]: What is this template for?

**Review by Zamm178 — APPROVED** [review]
Sorry for the delay in reviewing. I was able to build test this without issue and don't have any further comments on the code itself. Good work!

File: lib/tyk/configuration_templates/api_graphql.json.tpl (hunk 5 above)
Comment (AliRZ-02) [reply lib/tyk/configuration_templates/api_graphql.json.tpl:1]: This is part of the boilerplate code required for the Tyk API, to enable authentication and authorization for the GraphQL interface

**Review by shaikh-rashid — APPROVED** [review]
Great work Ali! LGTM 🎉

**Pull request merged**